### PR TITLE
Fix 'types' column in SHOW DATA CONNECTIONS [HZ-2384]

### LINKED ISSUE #24415: SHOW DATA CONNECTIONS prints 'java.util.ArrayList cannot be cast to com.hazelcast.core.HazelcastJsonValue' [HZ-2384]
**Describe the bug**
If any data connection is created then `SHOW DATA CONNECTIONS;` prints `java.util.ArrayList cannot be cast to com.hazelcast.core.HazelcastJsonValue`.

Failed in revision ebb6c97. It seems to be caused by any of latest commits (it worked fine ~2days ago).

**Expected behavior**
It should show data connections.

**To Reproduce**
```
sql> SHOW DATA CONNECTIONS;
+--------------------+----------------------------------------+
|name                |types                                   |
+--------------------+----------------------------------------+
+--------------------+----------------------------------------+
0 row(s) selected
sql> CREATE DATA CONNECTION mysql TYPE JDBC OPTIONS('jdbcUrl'='jdbc:mysql://localhost:3306/mydb', 'username'='root', 'password'='123');
OK
sql> SHOW DATA CONNECTIONS;
java.util.ArrayList cannot be cast to com.hazelcast.core.HazelcastJsonValue
```

### LINKED ISSUE #24421: com.hazelcast.jet.sql.impl.connector.jdbc.mysql.MySqlDownTest.showDataConnectionsShouldReturnQuickly [HZ-2384]
_master_ (commit 61f6bc146d77e3c5117ef62574becf258b43f4fe)

Failed on ibm-8: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-IbmJDK8-nightly/102/testReport/com.hazelcast.jet.sql.impl.connector.jdbc.mysql/MySqlDownTest/showDataConnectionsShouldReturnQuickly/
<details><summary>Stacktrace:</summary>

```
org.junit.ComparisonFailure: 

Expecting actual:
  [Row{[mysql, [Table]]}, Row{[testDatabaseRef, [Table]]}]
to contain exactly (and in same order):
  [Row{[mysql]}, Row{[testDatabaseRef]}]
but some elements were not found:
  [Row{[mysql]}, Row{[testDatabaseRef]}]
and others were not expected:
  [Row{[mysql, [Table]]}, Row{[testDatabaseRef, [Table]]}]
 expected:<[Row{[mysql[]}, Row{[testDatabaseRef]]}]> but was:<[Row{[mysql[, [Table]]}, Row{[testDatabaseRef, [Table]]]}]>
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:83)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:57)
	at com.hazelcast.jet.sql.impl.connector.jdbc.mysql.MySqlDownTest.showDataConnectionsShouldReturnQuickly(MySqlDownTest.java:93)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:825)

```
</details>

<details><summary>Standard output:</summary>

```
09:22:25,507  INFO |showDataConnectionsShouldReturnQuickly| - [0] Time-limited test - Creating container for image: ghcr.io/shopify/toxiproxy:2.5.0
09:22:25,508  INFO |showDataConnectionsShouldReturnQuickly| - [RegistryAuthLocator] Time-limited test - Failure when attempting to lookup auth config. Please ignore if you don't have images in an authenticated registry. Details: (dockerImageName: ghcr.io/shopify/toxiproxy:2.5.0, configFile: /home/jenkins/.docker/config.json, configEnv: DOCKER_AUTH_CONFIG). Falling back to docker-java default behaviour. Exception message: Status 404: No config supplied. Checked in order: /home/jenkins/.docker/config.json (file not found), DOCKER_AUTH_CONFIG (not set)
09:22:25,567 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:25,667 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:25,767 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:25,867 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:25,967 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:26,007  INFO |showDataConnectionsShouldReturnQuickly| - [0] Time-limited test - Container ghcr.io/shopify/toxiproxy:2.5.0 is starting: a59d71b4b601878a3fd04dd180b0a926ce65bba85940c2533eaa704a61738b7e
09:22:26,068 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:26,168 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:26,268 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:26,368 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-6 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:26,468 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:26,568 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:26,668 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:26,769 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:26,869 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:26,969 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:27,069 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:27,170 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:27,218  INFO |showDataConnectionsShouldReturnQuickly| - [HttpWaitStrategy] Time-limited test - /friendly_herschel: Waiting for 60 seconds for URL: http://172.17.0.1:54525/version (where port 54525 maps to container port 8474)
09:22:27,236  INFO |showDataConnectionsShouldReturnQuickly| - [0] Time-limited test - Container ghcr.io/shopify/toxiproxy:2.5.0 started in PT1.729S
09:22:27,270 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:27,370 DEBUG || - [JobCoordinationService] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
09:22:27,373  INFO |showDataConnectionsShouldReturnQuickly| - [PartitionStateManager] showDataConnectionsShouldReturnQuickly - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Initializing cluster partition table arrangement...
09:22:27,379  INFO |showDataConnectionsShouldReturnQuickly| - [HikariDataSource] showDataConnectionsShouldReturnQuickly - HikariPool-26-mysql - Starting...
09:22:27,397  INFO |showDataConnectionsShouldReturnQuickly| - [HikariDataSource] showDataConnectionsShouldReturnQuickly - HikariPool-26-mysql - Start completed.
09:22:27,473 DEBUG || - [JobRepository] hz.MySqlDownTest_youthful_yonath.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 0ms
09:22:29,350  INFO |showDataConnectionsShouldReturnQuickly| - [SqlTestSupport] Time-limited test - Removing 0 cached plans in SqlTestSupport.@After
09:22:29,350  INFO |showDataConnectionsShouldReturnQuickly| - [SqlTestSupport] Time-limited test - Removing 0 cached plans in SqlTestSupport.@After
09:22:29,351  INFO |showDataConnectionsShouldReturnQuickly| - [SimpleTestInClusterSupport] Time-limited test - Ditching 0 jobs in SimpleTestInClusterSupport.@After: []
09:22:29,352  INFO |showDataConnectionsShouldReturnQuickly| - [SimpleTestInClusterSupport] Time-limited test - Destroying 1 distributed objects in SimpleTestInClusterSupport.@After: [hz:impl:mapService/__sql.catalog]
BuildInfo right after showDataConnectionsShouldReturnQuickly(com.hazelcast.jet.sql.impl.connector.jdbc.mysql.MySqlDownTest): BuildInfo{version='5.3.0-SNAPSHOT', build='20230501', buildNumber=20230501, revision=61f6bc1, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'showDataConnectionsShouldReturnQuickly(com.hazelcast.jet.sql.impl.connector.jdbc.mysql.MySqlDownTest):'
09:22:25, accumulated pauses: 38 ms, max pause: 1 ms, pauses over 1000 ms: 0


No metrics recorded during the test

```
</details>

Standard output can be found here - https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-master-IbmJDK8-nightly/102/

### PR DESCRIPTION
**Pull request by burakgok**
Fixes #24415
Fixes #24421

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

**Comment by Fly-Style** [comment]
Also fixes https://github.com/hazelcast/hazelcast/issues/24421

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/DataConnectionResolver.java [hunk 1]
```diff
@@ -147,10 +152,18 @@ public static List<List<?>> getAllDataConnectionNameWithTypes(
...
 
         return conn.stream()
-                   .map(dc -> Arrays.asList(dc.getName(), new ArrayList<>(dc.resourceTypes())))
+                   .map(dc -> asList(dc.getName(), json(dc.resourceTypes())))
                    .collect(toList());
     }
 
+    private static HazelcastJsonValue json(Object value) {
+        try {
+            return new HazelcastJsonValue(SERIALIZER.writeValueAsString(value));
```
Comment (Fly-Style) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/DataConnectionResolver.java:161]: I have doubts regarding that part. Did you try `JsonArray` ?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlTestSupport.java [hunk 2]
```diff
@@ -455,7 +455,19 @@ protected static List<Row> allRows(SqlResult result) {
...
+
+    /**
+     * Execute a query and wait until it completes. Assert that the returned
+     * rows contain the expected rows, in the given order.
+     *
+     * @param instance     Hazelcast Instance to be used
+     * @param sql          The query
+     * @param expectedRows Expected rows
+     */
+    public static void assertRowsOrdered(HazelcastInstance instance, String sql, List<Row> expectedRows) {
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlTestSupport.java:469]: Great idea!

**Review by TomaszGaweda — APPROVED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/DataConnectionResolver.java (hunk 1 above)
Comment (burakgok) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/DataConnectionResolver.java]: We have three different JSON serializers: `JsonCreationUtil`, which uses Gson internally; Jackson `ObjectMapper`, which is used in tests; and finally `JsonValue.toString()`. I switched to the last one to get rid of external dependencies, but I don't know its performance characteristics.

**Comment by TomaszGaweda** [comment]
run-lab-run

**Review by Fly-Style — APPROVED** [review]

**Review by k-jamroz — CHANGES_REQUESTED** [review]
Update TDD with information about data returned by `SHOW DATA CONNECTIONS`

**Comment by k-jamroz** [comment]
As I said [somewhere else](https://github.com/hazelcast/hazelcast/issues/24415#issuecomment-1527220053), for `SHOW DATA CONNECTIONS` more important information is `CONNECTION TYPE`. `OBJECT TYPE` without connection type is almost meaningless. Possible `OBJECT TYPE`s make more sense in `SHOW RESOURCES`.

**Review by TomaszGaweda — APPROVED** [review]

**Review by TomaszGaweda — APPROVED** [review]

**Pull request merged**